src/remote/provisioning.rs

```diff
@@ -0,0 +1,1436 @@
+        self.write_xml(&mut vec).unwrap(); // safe
+        let xml = from_utf8(vec.as_slice()).unwrap(); // safe
+
+        xml.to_string()
+    }
+}
+
+/// # Decoding from XML
+/// 
+impl Message {
+    /// Parses an RFC 6492 <message />
+    pub fn decode<R: io::BufRead>(reader: R) -> Result<Self, Error> {
+        
+        let mut reader = xml::decode::Reader::new(reader);
+
+        let mut sender: Option<Sender> = None;
+        let mut recipient: Option<Recipient> = None;
+        let mut payload_type: Option<String> = None;
+
+        let mut outer = reader.start(|element| {
+            if element.name().local() != MESSAGE {
+                return Err(XmlError::Malformed)
+            }
+            
+            
+            element.attributes(|name, value| match name {
+                b"version" => {
+                    if value.ascii_into::<String>()? != VERSION {
+                        return Err(XmlError::Malformed)
+                    }
+                    Ok(())
+                }
+                b"sender" => {
+                    sender = Some(value.ascii_into()?);
+                    Ok(())
+                }
+                b"recipient" => {
+                    recipient = Some(value.ascii_into()?);
+                    Ok(())
+                }
+                b"type" => {
+                    payload_type = Some(value.ascii_into()?);
+                    Ok(())
+                }
+                _ => {
+                    Err(XmlError::Malformed)
+                }
+            })
+        })?;
+
+        // Get required attributes - return error if anything is missing.
+        let sender = sender.ok_or(XmlError::Malformed)?;
+        let recipient = recipient.ok_or(XmlError::Malformed)?;
+        let payload_type = payload_type.ok_or(XmlError::Malformed)?;
+
+        // Parse the nested payload
+        let payload = Payload::decode(
+            payload_type,
+            &mut outer,
+            &mut reader
+        )?;
+
+        // Check that there is no additional stuff
+        outer.take_end(&mut reader)?;
+        reader.end()?;
+
+        Ok(Message { sender, recipient, payload })
+    }
+
+}
+
+
+//------------ Payload -------------------------------------------------------
+
+/// Contains the query or reply payload of the message.
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[allow(clippy::large_enum_variant)]
+pub enum Payload {
+    List,
+    ListResponse(ResourceClassListResponse),
+    Issue(IssuanceRequest),
+    IssueResponse(IssuanceResponse),
+    Revoke(RevocationRequest),
+    RevokeResponse(RevocationResponse),
+    NotPerformedResponse(NotPerformedResponse),
+}
+
+/// # Decoding from XML
+/// 
+impl Payload {
+    /// Decodes the nested payload, needs to be given the value of the 'type'
+    /// attribute from the outer <message /> element so it can delegate to the
+    /// proper enclosed payload variant.
+    fn decode<R: io::BufRead>(
+        payload_type: String,
```

Comment:
Maybe have a special enum for the payload type with a `FromStr` impl for parsing it out of an attribute? That would save you an allocation for the happy path.

Reply:
done